lib/ecto/repo/schema.ex

```diff
@@ -590,8 +590,12 @@ defmodule Ecto.Repo.Schema do
         {:ok, values}
       {:invalid, _} = constraints ->
         constraints
-      {:error, :stale} ->
-        raise Ecto.StaleEntryError, struct: changeset.data, action: action
+      {:error, :no_change} ->
+        if changeset.filters != %{} do
+          raise Ecto.StaleEntryError, struct: changeset.data, action: action
+        else
+          {:ok, []}
+        end
```

Comment:
Instead of changing this here, which is used by insert/update/delete, I would probably change the `insert` operation to no longer call this apply function, as we want to handle `{:error, :stale}` differently.